stock_inventory_non_stocked/models/stock_inventory.py

```diff
@@ -0,0 +1,85 @@
+# Copyright 2024 Ivan Perez <iperez@coninpe.es>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+import ast
+
+from odoo import fields, models
+
+
+class StockInventory(models.Model):
+    _inherit = "stock.inventory"
+
+    create_non_stocked = fields.Boolean(
+        string="Create non stocked products",
+        help="If you select this option, a new quant with zero "
+        "quantity will be created for products without any stock and any quants",
+    )
+
+    def _get_quants(self, locations):
+        res = super(StockInventory, self)._get_quants(locations)
+        if self.create_non_stocked:
+            self._check_responsible_of_inventory()
+            # If the option selected is all, this domain will be applied
+            domain = [
+                ("has_quants", "=", False),
+                ("active", "=", True),
+                ("type", "=", "product"),
+            ]
+            if self.product_selection in ["manual", "one"]:
+                domain.append(("id", "in", self.product_ids.ids))
```

Comment:
A better writing is to use `osv.expression.AND`

Reply:
Done